{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyNwyBA0fhzRlS0LXZ1wOH90",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/farhanarnob/BioCounting/blob/main/GreedyMotifSearch%20with%20Pseudocounts.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 24,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "x8kp_X6ylrkN",
        "outputId": "510b6aa7-eb0f-4b30-af49-203e2a0ce9db"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "GCTTACGTATAC GCTTATGGAGGC GCTTATGAAACC GCTTAAGCATTC GCTTAAGAAACC GCTTACGAACTC GCTTATGAAAAC GCTTAGGTATCC GCTTAGGGAGTC GCTTAAGGAACC GCTTAAGGACTC GCTTAAGGAATC GCTTATGCATAC GCTTACGGATTC GCTTAAGTAGGC GCTTAAGCAGCC GCTTATGGAGAC GCTTACGGACAC GCTTATGAACGC GCTTAAGAAATC GCTTACGTAATC GCTTACGTAACC GCTTAGGAACAC GCTTAAGAAAGC GCTTAAGTAAGC\n"
          ]
        }
      ],
      "source": [
        "import sys\n",
        "import numpy as np \n",
        "nucleotides = [\"A\",\"C\",\"G\",\"T\"]\n",
        "def hammingDistance(stringOne, stringTwo):\n",
        "    count = 0\n",
        "    for i in range(0,len(stringOne)):\n",
        "        if stringOne[i]!=stringTwo[i]:\n",
        "            count += 1\n",
        "    return count\n",
        "\n",
        "def generator(s, k):\n",
        "    for i in range(1 + len(s) - k): yield s[i:i+k]\n",
        "\n",
        "def calProfile(profile, kmer):\n",
        "    probability = 1.0\n",
        "    for i in range (len(kmer)):\n",
        "        if kmer[i] == \"A\":\n",
        "            probability = probability * float(profile[i][0])\n",
        "        elif kmer[i] == \"C\":\n",
        "            probability = probability * float(profile[i][1])\n",
        "        elif kmer[i] == \"G\":\n",
        "            probability = probability * float(profile[i][2])\n",
        "        elif kmer[i] == \"T\":\n",
        "            probability = probability * float(profile[i][3])\n",
        "    return probability\n",
        "\n",
        "def mostProbableProfile(text,k,profile):\n",
        "    maxPrblity = 0.0\n",
        "    res = \"\"\n",
        "    for i in range(len(text)- k + 1):\n",
        "        kmer = text[i:i+k]\n",
        "        prblity = calProfile(profile, kmer)\n",
        "        if prblity > maxPrblity:\n",
        "            maxPrblity = prblity\n",
        "            res = kmer\n",
        "    return res \n",
        "\n",
        "def count(Motifs):\n",
        "    count = {}\n",
        "    k = len(Motifs[0])\n",
        "    for symbol in \"ACGT\":\n",
        "        count[symbol] = []\n",
        "        for i in range(k):\n",
        "            count[symbol].append(0)\n",
        "    t = len(Motifs)\n",
        "    for i in range(t):\n",
        "        for j in range(k):\n",
        "            symbol = Motifs[i][j]\n",
        "            count[symbol][j] += 1\n",
        "    return count\n",
        "\n",
        "def findConsensus(motifs):\n",
        "\n",
        "    consensus = \"\"\n",
        "    for i in range(len(motifs[0])):\n",
        "        countA, countC, countG, countT = 0, 0, 0, 0\n",
        "        for motif in motifs:\n",
        "            if motif[i] == nucleotides[0]:\n",
        "                countA += 1\n",
        "            elif motif[i] == nucleotides[1]:\n",
        "                countC += 1\n",
        "            elif motif[i] == nucleotides[2]:\n",
        "                countG += 1\n",
        "            elif motif[i] == nucleotides[3]:\n",
        "                countT += 1\n",
        "        if countA >= max(countC, countG, countT):\n",
        "            consensus += nucleotides[0]\n",
        "        elif countC >= max(countA, countG, countT):\n",
        "            consensus += nucleotides[1]\n",
        "        elif countG >= max(countC, countA, countT):\n",
        "            consensus += nucleotides[2]\n",
        "        elif countT >= max(countC, countG, countA):\n",
        "            consensus += nucleotides[3]\n",
        "    return consensus\n",
        "\n",
        "\n",
        "def profileMatrix(motifs):\n",
        "    Profile = {}\n",
        "    A, C, G, T = [], [], [], []\n",
        "    for j in range(len(motifs[0])):\n",
        "        countA, countC, countG, countT = 0, 0, 0, 0\n",
        "        for motif in motifs:\n",
        "            if motif[j] == nucleotides[0]:\n",
        "                countA += 1\n",
        "            elif motif[j] == nucleotides[1]:\n",
        "                countC += 1\n",
        "            elif motif[j] == nucleotides[2]:\n",
        "                countG += 1\n",
        "            elif motif[j] == nucleotides[3]:\n",
        "                countT += 1\n",
        "        A.append(countA)\n",
        "        C.append(countC)\n",
        "        G.append(countG)\n",
        "        T.append(countT)\n",
        "    Profile[\"A\"] = A\n",
        "    Profile[\"C\"] = C\n",
        "    Profile[\"G\"] = G\n",
        "    Profile[\"T\"] = T\n",
        "    return Profile\n",
        "\n",
        "\n",
        "def score(motifs):\n",
        "    consensus = findConsensus(motifs)\n",
        "    score = 0.0000\n",
        "    for motif in motifs:\n",
        "        score += hammingDistance(consensus, motif)\n",
        "    return round(score, 4)\n",
        "\n",
        "def greedyMotifSearch(DNA, k, t):\n",
        "    import math\n",
        "    bestMotifs = []\n",
        "    bestScore = math.inf\n",
        "    for string in DNA:\n",
        "        bestMotifs.append(string[:k])\n",
        "    base = DNA[0]\n",
        "    for i in generator(base, k):\n",
        "       \n",
        "        newMotifs = [i]\n",
        "        \n",
        "        for j in range(1, len(DNA)):\n",
        "           \n",
        "            profile = profileMatrix(newMotifs)\n",
        "            probable = mostProbableProfile(DNA[j], k, profile)\n",
        "            newMotifs.append(probable)\n",
        "\n",
        "        if score(newMotifs) < bestScore:\n",
        "       \n",
        "            bestScore = score(newMotifs)\n",
        "            bestMotifs = newMotifs\n",
        "    return bestMotifs\n",
        "\n",
        "\n",
        "\n",
        "def profilePseudocounts(motifs):\n",
        "\t'''Returns the profile of the dna list motifs.'''\n",
        "\tprof = []\n",
        "\tfor i in range(len(motifs[0])):\n",
        "\t\tcol = ''.join([motifs[j][i] for j in range(len(motifs))])\n",
        "\t\tprof.append([float(col.count(nuc)+1)/float(len(col)+4) for nuc in 'ACGT'])\n",
        "\treturn prof\n",
        "\n",
        "def greedyMotifSearchAlgorithm(Dna, k, t):\n",
        "    # Initialize the best score as a score higher than the highest possible score.\n",
        "    bestScore = k*t\n",
        "\n",
        "    # Run the greedy motif search.\n",
        "    for i in range(len(Dna[0])-k+1):\n",
        "        # Initialize the motifs as each k-mer from the first dna sequence.\n",
        "        motifs = [Dna[0][i:i+k]]\n",
        "\n",
        "        # Find the most probable k-mer in the next string.\n",
        "        for j in range(1, t):\n",
        "            currentProfile = profilePseudocounts(motifs)\n",
        "            motifs.append(mostProbableProfile(Dna[j], k, currentProfile))\n",
        "\n",
        "        # Check to see if we have a new best scoring list of motifs.\n",
        "        currentScore = score(motifs)\n",
        "        if currentScore < bestScore:\n",
        "            bestScore = currentScore\n",
        "            bestMotifs = motifs\n",
        "\n",
        "    return bestMotifs\n",
        "def main():\n",
        "    # reads input from terminal\n",
        "    input = \"\".join(open('/content/rosalind_ba2e.txt')).split()\n",
        "    # read input\n",
        "    k = int(input[0])\n",
        "    t = int(input[1])\n",
        "    dna = input[2:]\n",
        "\n",
        "    bestMotifs = greedyMotifSearchAlgorithm(dna, k, t)\n",
        "    print(*bestMotifs)\n",
        "\n",
        "\n",
        "if __name__ == \"__main__\":\n",
        "    main()"
      ]
    }
  ]
}