{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyPpPWIv2Ts0gu5gF/b2v1T/",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/farhanarnob/BioCounting/blob/main/RandomizedMotifSearch.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 23,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "vpb4Szq7t58A",
        "outputId": "92a81c5a-54bc-4205-a1f4-78189ebe62a7"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "TCAGTAAA\n",
            "TAAGTGCC\n",
            "ACCGAGAC\n",
            "TCGGTGAA\n",
            "CCAGCTCC\n"
          ]
        }
      ],
      "source": [
        "from random import randint\n",
        "from random import random\n",
        "import numpy as np\n",
        "nucleotides = [\"A\",\"C\",\"G\",\"T\"]\n",
        "def hammingDistance(stringOne, stringTwo):\n",
        "    count = 0\n",
        "    for i in range(0,len(stringOne)):\n",
        "        if stringOne[i]!=stringTwo[i]:\n",
        "            count += 1\n",
        "    return count\n",
        "\n",
        "def findConsensus(motifs):\n",
        "\n",
        "    consensus = \"\"\n",
        "    for i in range(len(motifs[0])):\n",
        "        countA, countC, countG, countT = 0, 0, 0, 0\n",
        "        for motif in motifs:\n",
        "            if motif[i] == nucleotides[0]:\n",
        "                countA += 1\n",
        "            elif motif[i] == nucleotides[1]:\n",
        "                countC += 1\n",
        "            elif motif[i] == nucleotides[2]:\n",
        "                countG += 1\n",
        "            elif motif[i] == nucleotides[3]:\n",
        "                countT += 1\n",
        "        if countA >= max(countC, countG, countT):\n",
        "            consensus += nucleotides[0]\n",
        "        elif countC >= max(countA, countG, countT):\n",
        "            consensus += nucleotides[1]\n",
        "        elif countG >= max(countC, countA, countT):\n",
        "            consensus += nucleotides[2]\n",
        "        elif countT >= max(countC, countG, countA):\n",
        "            consensus += nucleotides[3]\n",
        "    return consensus\n",
        "\n",
        "def profile(motif):\n",
        "    \n",
        "    t = len(motif)\n",
        "    k = len(motif[0])\n",
        "    profile = {0: [0]*k, 1: [0]*k, 2: [0]*k, 3: [0]*k}\n",
        "    # increment profile for each letter in mtif accordingly\n",
        "    for i in range(k):\n",
        "        for j in range(t):\n",
        "            if motif[j][i]== nucleotides[0]:\n",
        "                profile[0][i] +=1\n",
        "            elif motif[j][i]== nucleotides[1]:\n",
        "                profile[1][i] +=1\n",
        "            elif motif[j][i]== nucleotides[2]:\n",
        "                profile[2][i] +=1\n",
        "            elif motif[j][i]== nucleotides[3]:\n",
        "                profile[3][i] +=1\n",
        "    #Laplace's Rule of Sucession\n",
        "    for i in range(k):\n",
        "        for j in range(4):\n",
        "            profile[j][i] +=1\n",
        "    return profile\n",
        "\n",
        "\n",
        "def score(motifs):\n",
        "    consensus = findConsensus(motifs)\n",
        "    score = 0.0000\n",
        "    for motif in motifs:\n",
        "        score += hammingDistance(consensus, motif)\n",
        "    return round(score, 4)\n",
        "    \n",
        "    # sums up all HumminDistances for each motif. If letter is not the same as consensus string its being added to score fro all indexed of each motif\n",
        "    for motif in Motifs:\n",
        "        dist = np.zeros(len(total))\n",
        "        for i in range (len(total)):\n",
        "            if total[i]!= motif[i]:\n",
        "                score = score +1\n",
        "            \n",
        "    \n",
        "    return score\n",
        "\n",
        "def fMotifs(PROFILE,dna,k):\n",
        "    \n",
        "    newMotifs = []\n",
        "    \n",
        "    total = PROFILE[0][0] + PROFILE[1][0] + PROFILE[2][0] + PROFILE[3][0]\n",
        "    for genome in dna:\n",
        "        prob = []\n",
        "        motifs = []\n",
        "        for i in range(len(genome)-k+1):\n",
        "            motif = genome[i:i+k]\n",
        "            motifs.append(motif)\n",
        "            propability = 1.0\n",
        "            for j in range(len(motif)):\n",
        "                if motif[j] == nucleotides[0]:\n",
        "                    val = 0\n",
        "                elif motif[j] == nucleotides[1]:\n",
        "                    val = 1\n",
        "                elif motif[j] == nucleotides[2]:\n",
        "                    val = 2\n",
        "                elif motif[j] == nucleotides[3]:\n",
        "                    val = 3\n",
        "                propability = propability *(float(PROFILE[val][j]) / total )\n",
        "            prob.append(propability)\n",
        "        \n",
        "        prob = np.array(prob)\n",
        "        idx = np.where(prob == prob.max())\n",
        "        idx = idx[0][0]\n",
        "        newMotifs.append(motifs[idx])\n",
        "    \n",
        "    return newMotifs\n",
        "    \n",
        "def randomizedMotifSearch(dna,k,t):\n",
        "    bestmotifs = []\n",
        "    molifs = []\n",
        "    for motif in dna:\n",
        "        rand = randint(0, len(motif)-k)\n",
        "        newMotif = motif[rand:rand+k]\n",
        "        molifs.append(newMotif)\n",
        "    bestmotifs = molifs\n",
        "    \n",
        "    while (1==1):\n",
        "        profileScore = profile(molifs)\n",
        "        \n",
        "        \n",
        "        motifs = fMotifs(profileScore, dna,k)\n",
        "        if score(molifs) < score (bestmotifs):\n",
        "            bestmotifs = molifs\n",
        "        else:\n",
        "            return bestmotifs\n",
        "        \n",
        "    return bestmotifs\n",
        "def main():\n",
        "    # reads input from terminal\n",
        "    input = \"\".join(open('/content/input_1.txt')).split()\n",
        "    # read input\n",
        "    k = int(input[0])\n",
        "    d = int(input[1])\n",
        "    dna = input[2:]\n",
        "\n",
        "    min = 1e6\n",
        "    for i in range(1000):\n",
        "        res = randomizedMotifSearch(dna, k, d)\n",
        "        #compare the scores from each random start and select the lowest for motifs produced\n",
        "        resScore = score(res)\n",
        "        if resScore < min:\n",
        "            min = resScore\n",
        "            result = res\n",
        "    \n",
        "    \n",
        "    print(\"\\n\".join(result))\n",
        " \n",
        "\n",
        "if __name__ == \"__main__\":\n",
        "    main()"
      ]
    }
  ]
}