{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyOCkMC3RpCZhrIizUrYSkAq",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/farhanarnob/BioCounting/blob/main/RandomizedMotifSearch.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "iLzbrv3py4mz",
        "outputId": "04b83921-3da0-42b9-d23e-6edc1c3dc7e8"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "TCGTACTCCCGTTAG\n",
            "TATGCTTCTAGTTCG\n",
            "TATCACTCTAGTGGA\n",
            "TATATATCTAGTTCG\n",
            "TATCAAATTAGTTCG\n",
            "TATCACCTAAGTTCG\n",
            "TATCACTCTTTATCG\n",
            "TATCACTCTAGACGG\n",
            "TATCGAGCTAGTTCG\n",
            "TATCACATGAGTTCG\n",
            "TACTGCTCTAGTTCG\n",
            "CCTCACTCTAGTTCC\n",
            "AGCCACTCTAGTTCG\n",
            "TATCAGGTTAGTTCG\n",
            "TATCGTCCTAGTTCG\n",
            "AATCACTCTAGTTAA\n",
            "TATCACTCTAACACG\n",
            "TATCACTCCGCTTCG\n",
            "TGCAACTCTAGTTCG\n",
            "TATCACTTATGTTCG\n"
          ]
        }
      ],
      "source": [
        "from random import randint\n",
        "from random import random\n",
        "import numpy as np\n",
        "\n",
        "\n",
        "nucleotides = [\"A\",\"C\",\"G\",\"T\"]\n",
        "#RandomizedMotifSearch\n",
        "def PROFILE(motif):\n",
        "    \n",
        "    t = len(motif)\n",
        "    k = len(motif[0])\n",
        "    profile = {0: [0]*k, 1: [0]*k, 2: [0]*k, 3: [0]*k}\n",
        "    # increment profile for each letter in mtif accordingly\n",
        "    for i in range(k):\n",
        "        for j in range(t):\n",
        "            if motif[j][i]== nucleotides[0]:\n",
        "                profile[0][i] +=1\n",
        "            elif motif[j][i]== nucleotides[1]:\n",
        "                profile[1][i] +=1\n",
        "            elif motif[j][i]== nucleotides[2]:\n",
        "                profile[2][i] +=1\n",
        "            elif motif[j][i]== nucleotides[3]:\n",
        "                profile[3][i] +=1\n",
        "    #Laplace's Rule of Sucession\n",
        "    for i in range(k):\n",
        "        for j in range(4):\n",
        "            profile[j][i] +=1\n",
        "    return profile\n",
        "\n",
        "\n",
        "   \n",
        "def SCORE(Motifs):\n",
        "    total = ''\n",
        "    # created PROFILE for given Motifs\n",
        "   \n",
        "    profile = PROFILE(Motifs)\n",
        "    k = len(Motifs[0])\n",
        "    # Create the consensus string by checking which ATCG is the most frequent in the PROFILE for each i\n",
        "    for i in range(k):\n",
        "        biggest = 0\n",
        "        for j in [0, 1, 2, 3]:\n",
        "            if profile[j][i] > biggest:\n",
        "                biggest = profile[j][i]\n",
        "                if j==0: \n",
        "                    a = nucleotides[0]\n",
        "                elif j==1:\n",
        "                    a = nucleotides[1]\n",
        "                elif j ==2:\n",
        "                    a = nucleotides[2]\n",
        "                elif j == 3:\n",
        "                    a = nucleotides[3]\n",
        "        total += a\n",
        "    score = 0\n",
        "    \n",
        "    # sums up all HumminDistances for each motif. If letter is not the same as consensus string its being added to score fro all indexed of each motif\n",
        "    for motif in Motifs:\n",
        "        dist = np.zeros(len(total))\n",
        "        for i in range (len(total)):\n",
        "            if total[i]!= motif[i]:\n",
        "                score = score +1\n",
        "            \n",
        "    \n",
        "    return score\n",
        "def MOTIFS(PROFILE,dna,k):\n",
        "    \n",
        "    newMotifs = []\n",
        "    \n",
        "    total = PROFILE[0][0] + PROFILE[1][0] + PROFILE[2][0] + PROFILE[3][0]\n",
        "    for genome in dna:\n",
        "        prob = []\n",
        "        motifs = []\n",
        "        for i in range(len(genome)-k+1):\n",
        "            motif = genome[i:i+k]\n",
        "            motifs.append(motif)\n",
        "            propability = 1.0\n",
        "            for j in range(len(motif)):\n",
        "                if motif[j] == nucleotides[0]:\n",
        "                    val = 0\n",
        "                elif motif[j] == nucleotides[1]:\n",
        "                    val = 1\n",
        "                elif motif[j] == nucleotides[2]:\n",
        "                    val = 2\n",
        "                elif motif[j] == nucleotides[3]:\n",
        "                    val = 3\n",
        "                propability = propability *(float(PROFILE[val][j]) / total )\n",
        "            prob.append(propability)\n",
        "        \n",
        "        prob = np.array(prob)\n",
        "        idx = np.where(prob == prob.max())\n",
        "        idx = idx[0][0]\n",
        "        newMotifs.append(motifs[idx])\n",
        "    \n",
        "    return newMotifs\n",
        "    \n",
        "def randomizedMotifSearch(dna,k,t):\n",
        "    Bestmotifs = []\n",
        "    Motifs = []\n",
        "    for motif in dna:\n",
        "        rand = randint(0, len(motif)-k)\n",
        "        newMotif = motif[rand:rand+k]\n",
        "        Motifs.append(newMotif)\n",
        "    Bestmotifs = Motifs\n",
        "    \n",
        "    while (1==1):\n",
        "        profile = PROFILE(Motifs)\n",
        "        \n",
        "        \n",
        "        Motifs = MOTIFS(profile, dna,k)\n",
        "        if SCORE(Motifs) < SCORE (Bestmotifs):\n",
        "            Bestmotifs = Motifs\n",
        "        else:\n",
        "            return Bestmotifs\n",
        "        \n",
        "    return Bestmotifs\n",
        "def main():\n",
        "    # reads input from terminal\n",
        "    input = \"\".join(open('/content/rosalind_ba2f (1).txt')).split()\n",
        "    # read input\n",
        "    k = int(input[0])\n",
        "    d = int(input[1])\n",
        "    dna = input[2:]\n",
        "\n",
        "    min = 1e6\n",
        "    for i in range(1000):\n",
        "        res = randomizedMotifSearch(dna, k, d)\n",
        "        #compare the scores from each random start and select the lowest for motifs produced\n",
        "        resScore = SCORE(res)\n",
        "        if resScore < min:\n",
        "            min = resScore\n",
        "            result = res\n",
        "    \n",
        "    \n",
        "    print(\"\\n\".join(result))\n",
        " \n",
        "\n",
        "if __name__ == \"__main__\":\n",
        "    main()"
      ]
    }
  ]
}